File: src/main/java/linq/collections/enumerator/IterableEnumerator.java
package linq.collections.enumerator;

import linq.Enumerator;

import java.util.Iterator;

public class IterableEnumerator<T> implements Enumerator<T> {
    private final Iterator<T> iterator;
    private T current = null;

    public IterableEnumerator(Iterator<T> iterator) {
        this.iterator = iterator;
    }

    @Override
    public boolean moveNext() {
        if (iterator.hasNext()) {
            current = iterator.next();
            return true;
        }
        return false;
    }

    @Override
    public T current() {
        return current;
    }
}

File: src/main/kotlin/linq/collections/enumerator/IterableEnumerator.kt
package linq.collections.enumerator

import linq.Enumerator

class IterableEnumerator<T>(iterable: Iterable<T>) : Enumerator<T> {
    private val iterator: Iterator<T> = iterable.iterator()
    private var backingCurrent: T? = null

    override val current: T
        get() = backingCurrent ?: throw IllegalStateException()


    @Override
    override fun moveNext(): Boolean {
        if (iterator.hasNext()) {
            backingCurrent = iterator.next()
            return true
        }
        return false
    }
}
